## Air Quality 

**Questions**
<br>1. Create a new dataframe that contains all the values of the original dataframe but with no missing values. How
many rows does this dataframe contain?
<br>2. Create a subset of the original dataframe that contains only Ozone levels less than 31 and Temperature more
than 70. How many rows of this subset dataframe?
<br>3. Create a new dataframe that contains only the columns, Solar.R, Temp, and Month and arrange its rows by the
temperature value in a decreasing order. Which month is in the first rows of this dataframe?
<br>4. Which month has the highest mean temperature? Which month has the highest mean Solar radiation? Are
these two months the same month?
<br>5. All these airquality measurements were taken in the year 1973. Which weekday has the highest mean
temperature? (Hint: create a new variable “date” with the day, month, and year=1973 then pull the name of
the weekday from this variable) 

In [1]:
import pandas as pd

In [13]:
raw=pd.read_csv("./Assign 4/airquality.csv")

In [14]:
raw.head()

Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5

#### 1. Create a new dataframe that contains all the values of the original dataframe but with no missing values. How many rows does this dataframe contain? 

In [15]:
len(raw.index)

153

In [21]:
air=raw.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)

In [22]:
len(air.index)

111

There are 111 rows after removing null values

#### 2. Create a subset of the original dataframe that contains only Ozone levels less than 31 and Temperature more than 70. How many rows of this subset dataframe?

In [34]:
air2 =raw[(raw["Ozone"]<31) & (raw["Temp"]>70)]

In [35]:
len(air2.index)

33

There are 33 rows in subset dataframe

#### 3.  Create a new dataframe that contains only the columns, Solar.R, Temp, and Month and arrange its rows by the temperature value in a decreasing order. Which month is in the first rows of this dataframe? 

In [46]:
air3=raw[["Solar.R","Temp","Month"]]

In [48]:
air3.head()

Solar.R  Temp  Month
0    190.0    67      5
1    118.0    72      5
2    149.0    74      5
3    313.0    62      5
4      NaN    56      5

In [50]:
air3.sort_values("Temp",ascending=False,inplace=True)

C:\Users\sushrut.t\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [52]:
air3.head()

Solar.R  Temp  Month
119    203.0    97      8
121    237.0    96      8
122    188.0    94      8
120    225.0    94      8
41     259.0    93      6

The month in the first row is **8**

#### 4.Which month has the highest mean temperature? Which month has the highest mean Solar radiation? Are these two months the same month? 

In [54]:
# Mean of temperature & Solar radiation of each month
raw.groupby("Month").agg("mean")

Ozone     Solar.R       Wind       Temp   Day
Month                                                   
5      23.615385  181.296296  11.622581  65.548387  16.0
6      29.444444  190.166667  10.266667  79.100000  15.5
7      59.115385  216.483871   8.941935  83.903226  16.0
8      59.961538  171.857143   8.793548  83.967742  16.0
9      31.448276  167.433333  10.180000  76.900000  15.5

Month 7 has highet mean Solar radiation and Month 8 has highest mean temperature. 
<br>Even though month 7 & 8 have highest mean temperatures, month 8 has slighlty higher mean temperature. 

#### 5 All these airquality measurements were taken in the year 1973. Which weekday has the highest mean temperature? (Hint: create a new variable “date” with the day, month, and year=1973 then pull the name of the weekday from this variable)

In [57]:
raw["date"]=pd.to_datetime(1973*10000+raw.Month*100+raw.Day,format='%Y%m%d')

In [62]:
raw.head()

Ozone  Solar.R  Wind  Temp  Month  Day       date
0   41.0    190.0   7.4    67      5    1 1973-05-01
1   36.0    118.0   8.0    72      5    2 1973-05-02
2   12.0    149.0  12.6    74      5    3 1973-05-03
3   18.0    313.0  11.5    62      5    4 1973-05-04
4    NaN      NaN  14.3    56      5    5 1973-05-05

In [63]:
raw["weekday"]=raw['date'].dt.day_name()

In [64]:
raw.head()

Ozone  Solar.R  Wind  Temp  Month  Day       date    weekday
0   41.0    190.0   7.4    67      5    1 1973-05-01    Tuesday
1   36.0    118.0   8.0    72      5    2 1973-05-02  Wednesday
2   12.0    149.0  12.6    74      5    3 1973-05-03   Thursday
3   18.0    313.0  11.5    62      5    4 1973-05-04     Friday
4    NaN      NaN  14.3    56      5    5 1973-05-05   Saturday

In [68]:
raw.groupby("weekday").agg("mean")

Ozone     Solar.R       Wind       Temp     Month        Day
weekday                                                                    
Friday     31.600000  174.952381  10.604545  77.181818  7.000000  16.090909
Monday     37.437500  193.500000   8.947619  77.904762  7.000000  15.619048
Saturday   50.333333  188.150000   9.581818  77.636364  7.045455  15.681818
Sunday     44.388889  222.473684  10.031818  77.500000  7.090909  15.318182
Thursday   38.625000  177.500000  10.022727  78.590909  6.954545  16.500000
Tuesday    46.473684  191.909091  10.600000  79.136364  6.909091  15.909091
Wednesday  44.647059  158.409091   9.868182  77.227273  6.954545  15.500000

Tuesday has highest mean temperature **(79.136)**